File: golang/barbershop.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"

	"github.com/pkg/profile"
)

var chairs = 5
var customers = 100
var wg sync.WaitGroup

func sleepThread() {
	time.Sleep(time.Duration(rand.Intn(250)) * time.Millisecond)
}

func main() {
	defer profile.Start().Stop()
	//Buffered to number of chairs
	chairs := make(chan int, chairs)
	//Blocks on both ends untill barber and customer are both ready
	barberAvailable := make(chan int)
	//Start the barber
	go cutHair(chairs, barberAvailable)

	//Start the customers at random times
	for i := 0; i < customers; i++ {
		wg.Add(1)
		sleepThread()
		go getHairCut(chairs, barberAvailable, i)
	}
	//Make sure we wait till all the customers have visited the shop
	wg.Wait()
}

func cutHair(chairs, barberAvailable chan int) {
	for {
		//Wait till someone enters
		barberAvailable <- 1
		//Free up a chair
		<-chairs
		//cut hair
		sleepThread()
	}
}

func getHairCut(chairs, barberAvailable chan int, id int) {
	defer wg.Done()
	fmt.Printf("Customer %d: Arriving to barbershop\n", id)
	//Check if barber is available
	select {
	case <-barberAvailable:
		fmt.Printf("Customer %d: Getting haircut\n", id)
		return
	default:

	}

	select {
	case chairs <- 1:
		//Sit in chair
		fmt.Printf("Customer %d: Waiting in chair\n", id)
		//Wait till barber is available
		<-barberAvailable
		fmt.Printf("Customer %d: Getting haircut\n", id)
		return
	default:
		fmt.Printf("Customer %d: All chairs taken, leaving\n", id)
	}
}

File: java/Barbershop.java
import java.util.Random;
import java.util.concurrent.Semaphore;

//Solution taken from little book of semaphores and applied to java
public class Barbershop {

    static Semaphore mutex = new Semaphore(1);
    static Semaphore customerMutex = new Semaphore(0);
    static Semaphore barberMutex = new Semaphore(0);
    static Semaphore customerDoneMutex = new Semaphore(0);
    static Semaphore barberDoneMutex = new Semaphore(0);
    static int numReaders;
    static int valueToRead;
    static int chairs = 5;
    static int currentCustomers = 0;
    static int numCustomers = 100;

    public static void sleepThread() {
        try {
            Thread.sleep((long) (Math.random() * 250));
        } catch (InterruptedException e) {
            System.out.println(e.getMessage());
        }
    }

    public static void main(String args[]) {
        Customer customer = new Customer();
        Barber barber = new Barber();

        // init barber
        Thread barberThread = new Thread(barber);
        barberThread.setName("barber");
        // Quit program once customers are done
        barberThread.setDaemon(true);
        barberThread.start();

        Thread[] threads = new Thread[numCustomers];
        for (int i = 0; i < threads.length; i++) {
            sleepThread();
            threads[i] = new Thread(customer);
            threads[i].setName("Customer " + i);
            threads[i].start();
        }
    }

    static class Customer implements Runnable {
        @Override
        public void run() {
            try {
                Random rand = new Random();
                System.out.println(Thread.currentThread().getName() + ": Arriving to barbershop");
                // Make sure no other reader is trying to add itself
                mutex.acquire();
                if (currentCustomers == chairs) {
                    mutex.release();
                    System.out.println(Thread.currentThread().getName() + ": All chairs taken, leaving");
                    return;
                }
                // Increment current customers to get a haircut next
                currentCustomers++;
                mutex.release();

                customerMutex.release();
                // Wait for barber to be available
                System.out.println(Thread.currentThread().getName() + ": Waiting in chair");
                barberMutex.acquire();

                // Get hair cut
                System.out.println(Thread.currentThread().getName() + ": Getting hair cut");

                customerDoneMutex.release();
                barberDoneMutex.acquire();

                // Decrement current customers
                mutex.acquire();
                currentCustomers--;
                mutex.release();

            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    static class Barber implements Runnable {
        @Override
        public void run() {
            while (true) {
                try {
                    // Make sure no customers are already being cut
                    customerMutex.acquire();
                    barberMutex.release();

                    // Cut hair
                    sleepThread();

                    // release and wait for next customer
                    customerDoneMutex.acquire();
                    barberDoneMutex.release();
                } catch (InterruptedException e) {
                    System.out.println(e.getMessage());
                }
            }
        }
    }
}
